program-runtime/src/loaded_programs.rs

```diff
@@ -610,6 +605,15 @@ pub struct ProgramCache<FG: ForkGraph> {
     ///
     /// The first level is for the address at which programs are deployed and the second level for the slot (and thus also fork).
     entries: HashMap<Pubkey, SecondLevel>,
+    /// The entries that are getting loaded and have not yet finished loading.
+    ///
+    /// The key is the program address, the value is a tuple of the slot in which the program is
+    /// being loaded and the thread ID doing the load.
+    ///
+    /// It is possible that multiple TX batches from different slots need different versions of a
+    /// program. The deployment slot of a program is only known after load tho,
+    /// so all loads for a given program key are serialized.
+    loading_entries: Mutex<HashMap<Pubkey, (Slot, std::thread::ThreadId)>>,
```

Comment:
In the replacement of the index structure I am working on I also experimented with pulling `cooperative_loading_lock` out of the `SecondLevel`. We might be able to take this even further because when we use the account last-modified-slot we could also parallelize the loading of missing programs. Basically the "The deployment slot of a program is only known after load tho" part could be improved then.